lib/storage.js

```diff
@@ -58,23 +58,29 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
     // enable queries by controller and reference ID (reference IDs are
     // scoped to controllers)
     collection: 'zcap-storage',
-    fields: {controller: 1, referenceId: 1},
+    fields: {referenceId: 1, controller: 1},
```

Comment:
This change doesn't seem correct -- `referenceId` is meant to be scoped to `controller`. We'll want to discuss this further wrt. "the wrong indexes were being used".

Reply:
@dlongley Well the same structure was used with `zcap-authorization` where `id` is a required field and the other fields are optional. 

The problem here is that since `controller` is required and the first index, it will always use `{controller: 1, referenceId: 1}`(if change is not made) even if `referenceId` is not provided. For example, in the case of `id` being provided, it will still use the index I mentioned instead of `{controller: 1, id: 1}` because it is only indexing on `controller`. If that is what is intentional here, then we do not need these compound indexes since they are not used at all.

Reply:
Also in `get()` and `remove()` you must provide a secondary optional field of `id` or `referenceId`. This means that you cannot do an indexed query just based on `controller` anyways. However `find()` does not have that limitation, so it is possible there to query only be `controller`. However either way we need some correction here because we currently have unused indexes of `{controller: 1, id: 1}` and `{controller: 1, invoker: 1}`.

Reply:
@JoshDunnDev, it sounds like we may need to set aside some time to chat about this over voice. One thing I will ask, however, is: have you considered that some of these indexes are required to enforce uniqueness constraints? That is, they are not just used for queries, but they ensure that you can't have certain duplicates that should be considered illegal.